### PR TITLE
[batch-beta] Feature/batch application templates

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
The content to be added to Changelog is as follows:
* Application Template support for xplat cli

### CODE REVIEW

**Comment by theunrepentantgeek** [comment]
Thanks for fixing the title @amarzavery.

File: Documentation/BatchDocumentation/application-templates.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+| displayName             | String       | A display name for the job.                                                                                                                                                                                                                                  |
+| priority                | Int32        | The priority of the job. <br/> Prioritization of jobs is the responsibility of the end user, not an application template author.                                                                                                                             |
+| constraints             | Complex Type | Specifies the execution constraints for the job. <br/>The selection of appropriate constraints requires knowledge of the specific job being processed (e.g. size) and associated pool                                                                        |
+| poolInfo                | Complex Type | Specifies the pool on which the Batch service runs the job’s tasks. <br/> Establishing and managing a pool is the responsibility of the end user, not an application template author, not least because a forgotten pool might result in an unexpected cost. |
```
Comment (davefellows) [inline Documentation/BatchDocumentation/application-templates.md]: minor; suggest "...result in unexpected costs."

File: Documentation/BatchDocumentation/application-templates.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+The `applicationTemplateInfo` element gives the `filePath` to the application template and provides all the `parameters` required to configure the application template for use.
+
+## Samples
+
+[**Blender**](samples/blender-appTemplate/readme.md) - using an application template to specify the rendering of a blender scene. This sample uses the experimental [Task Factories](taskFactories.md) feature. 
+
+
+## Job schema changes
+
+The batch job gains the following element, used to specify which application template should be used:
```
Comment (jasper-schneider) [inline Documentation/BatchDocumentation/application-templates.md]: >batch [](start = 4, length = 5)

capitalize to Batch for consistency with our other docs?

File: Documentation/BatchDocumentation/application-templates.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+These newly introduced properties are used to define the templating capabilities.
+
+| Element name     | Required | Type         | Description                                                                                                                                                       |
+| ---------------- | -------- | ------------ | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| templateMetadata | Optional | Complex Type | Additional information about the template itself. <br/>Included for documentation purposes only. None of this information is passed through to the generated job. |
+| parameters       | Optional | Dictionary   | A (potentially empty) dictionary of parameter definitions, indexed by the name of the property.                                                                   |
+
+`templateMetadata`
+
+Though completely optional and not actually used by the xplatcli, template metadata is supported to allow the templates to be somewhat self documenting. These properties are recommended for each template.
```
Comment (jasper-schneider) [inline Documentation/BatchDocumentation/application-templates.md]: >xplatcli [](start = 56, length = 8)

Based on the repo's top level README, it looks like they use the branding "Xplat-CLI"

File: Documentation/BatchDocumentation/samples/blender-appTemplate/pool.json [hunk 4]
```diff
@@ -0,0 +1,57 @@
+{
+    "parameters": {
+        "vmSize": {
+            "type": "string",
+            "defaultValue": "STANDARD_D1",
+            "metadata": {
+                "description": "The size of the virtual machines that runs the application"
```
Comment (jasper-schneider) [inline Documentation/BatchDocumentation/samples/blender-appTemplate/pool.json]: >runs [](start = 70, length = 4)

"run" to match "machines", right?

File: Documentation/BatchDocumentation/samples/blender-appTemplate/readme.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+This template shows how to use `blender` to render using an *application template* to separate the logic of processing from adminstration and management.
+
+With an *application template*, the processing steps required for the job are defined in a separate file - see `render-template.json` which is appropriately parameterized. The job itself references the template, supplies any required parameter values and specifies the pool on which the job is to run.
+
+## Prerequisites
+
+You must have an Azure Batch account set up with a linked Azure Storage account.
+
+You will need a blender scene to render.
```
Comment (jasper-schneider) [inline Documentation/BatchDocumentation/samples/blender-appTemplate/readme.md]: >blender [](start = 16, length = 7)

same capitalization comment for this file

File: lib/commands/batch/batch.batchShipyardUtils._js [hunk 6]
```diff
@@ -255,12 +255,12 @@ batchShipyardUtils.createConfigJsonFromPool = function (poolJson) {
       dockerRegistry['login'] = login;
     }
     if (!__.isUndefined(poolJson.clientExtensions.dockerOptions.registry.private)) {
-      var private = {};
+      var privateOptions = {};
```
Comment (jasper-schneider) [inline lib/commands/batch/batch.batchShipyardUtils._js]: >privateOptions [](start = 10, length = 14)

thanks

File: lib/commands/batch/batch.templateUtils._js [hunk 7]
```diff
@@ -695,9 +696,9 @@ function parseArmConcat(content, template, parameters) {
  * Determine if a section of the template is an ARM reference, and calculate
  * the replacement accordingly. The result will be correctly typed to suit the 
  * parameter definition (e.g. will return a number if the parameter requires a number)
- * @param {string} value            a section of the template contained within []
- * @param {object} jsonTemplate     contents of the template file as an object
- * @param {object} properties       contents of the parameters file as an object
+ * @param {string} value - a section of the template contained within []
```
Comment (jasper-schneider) [inline lib/commands/batch/batch.templateUtils._js]: > -  [](start = 24, length = 3)

So I have no problem with adding hyphens here, but I believe that basically every one of our files does not and instead use some (inconsistent) amount of whitespace.
Just wondering why you're adding the hyphens only here but leaving the other files you touched unchanged?

File: lib/commands/batch/batch.templateUtils._js [hunk 8]
```diff
@@ -876,14 +877,275 @@ batchTemplateUtils.parseTemplate = function (template, jsonTemplate, parameters)
...
+};
+
+/**
+ * Validate the provided metadata is valid
+ */
+batchTemplateUtils.validateMetadata = function(metadata) {
+  // Rule: The prefix az_batch is reserved for our use and can't be specified on job nor template
+  var violation = metadata.find( function(item) { return item.name.startsWith('az_batch:'); });
+  if (!__.isUndefined(violation)) {
+    throw new Error(util.format('Metadata item \'%s\' cannot be used as the prefix \'az_batch:\' is reserved for batch use.', violation.name));
```
Comment (jasper-schneider) [inline lib/commands/batch/batch.templateUtils._js]: >batch [](start = 113, length = 5)

capitalize to Batch

File: test/data/batch-applicationTemplate-prohibitedPoolInfo.json [hunk 9]
```diff
@@ -0,0 +1,10 @@
+{
+  "templateInfo": {
+    "description": "A test application template that specifies prohibited property 'poolInfo'."
+  },
+  "poolInfo": {
+      "poolId" : "swimming"
```
Comment (jasper-schneider) [inline test/data/batch-applicationTemplate-prohibitedPoolInfo.json]: >swimming [](start = 18, length = 8)

I love cheesy jokes as much as the next guy, but this one's pushing it :)

File: Documentation/BatchDocumentation/application-templates.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+All of the details about how the job should be processed are moved into the **application template**, creating a reusable definition that is independent of a particular account. Application templates are parameterized to allow the processing to be customized without requiring modification of the template itself.
+
+The job itself retains all of the account specific configuration for the job, specifying the appropriate pool, any runtime constraints and so on.
+
+To link the two, the job now references the required application template and supplies any parameters required to customize processing for the needs of the current user.
+
+## Example job
+
+This sample blender job specifies rendering of a blender scene using application templates.
```
Comment (theunrepentantgeek) [inline Documentation/BatchDocumentation/application-templates.md]: Will fix.

File: Documentation/BatchDocumentation/application-templates.md (hunk 1 above)
Comment (theunrepentantgeek) [inline Documentation/BatchDocumentation/application-templates.md]: Will amend.

File: Documentation/BatchDocumentation/samples/blender-appTemplate/pool.json (hunk 4 above)
Comment (theunrepentantgeek) [inline Documentation/BatchDocumentation/samples/blender-appTemplate/pool.json]: Yes, will fix.

File: Documentation/BatchDocumentation/samples/blender-appTemplate/readme.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
+# Azure Batch Blender using Application Templates
+
+This template shows how to use `blender` to render using an *application template* to separate the logic of processing from adminstration and management.
```
Comment (theunrepentantgeek) [inline Documentation/BatchDocumentation/samples/blender-appTemplate/readme.md]: Will fix.

File: Documentation/BatchDocumentation/samples/blender-appTemplate/readme.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+With an *application template*, the processing steps required for the job are defined in a separate file - see `render-template.json` which is appropriately parameterized. The job itself references the template, supplies any required parameter values and specifies the pool on which the job is to run.
+
+## Prerequisites
+
+You must have an Azure Batch account set up with a linked Azure Storage account.
```
Comment (theunrepentantgeek) [inline Documentation/BatchDocumentation/samples/blender-appTemplate/readme.md]: Found an up to date link, included.

File: Documentation/BatchDocumentation/samples/blender-appTemplate/readme.md (hunk 5 above)
Comment (theunrepentantgeek) [inline Documentation/BatchDocumentation/samples/blender-appTemplate/readme.md]: Fixed.

File: Documentation/BatchDocumentation/samples/blender-appTemplate/render-template.json [hunk 13]
```diff
@@ -0,0 +1,79 @@
+{
+    "templateMetadata" : {
+        "description" : "Sample application template for working with blender."
```
Comment (theunrepentantgeek) [inline Documentation/BatchDocumentation/samples/blender-appTemplate/render-template.json]: Fixed.

File: lib/commands/batch/batch.templateUtils._js [hunk 14]
```diff
@@ -381,8 +382,8 @@ function isSubstitution(template, start, end) {
 
 /**
  * Replace string with sweep value
- * @param {array} data            The string to be replaced
- * @param {array} parameters      The sweep value, each value map to {0}, {1}, .. {n} 
+ * @param {strung} data - The string to be replaced
```
Comment (theunrepentantgeek) [inline lib/commands/batch/batch.templateUtils._js]: Fixed.

File: lib/commands/batch/batch.templateUtils._js [hunk 15]
```diff
@@ -381,8 +382,8 @@ function isSubstitution(template, start, end) {
...
+ * @param {array} parameters - The sweep value, each value map to {0}, {1}, .. {n}
```
Comment (theunrepentantgeek) [inline lib/commands/batch/batch.templateUtils._js]: Updated wording in comment.

File: test/data/batch-applicationTemplate-prohibitedPoolInfo.json (hunk 9 above)
Comment (theunrepentantgeek) [inline test/data/batch-applicationTemplate-prohibitedPoolInfo.json]: Thanks for noticing.

File: lib/commands/batch/batch.templateUtils._js (hunk 8 above)
Comment (theunrepentantgeek) [inline lib/commands/batch/batch.templateUtils._js]: Reworded to remove ambiguity - `az_batch:` is the prefix, not the metadata item.

File: Documentation/BatchDocumentation/samples/blender-appTemplate/readme.md (hunk 11 above)
Comment (theunrepentantgeek) [inline Documentation/BatchDocumentation/samples/blender-appTemplate/readme.md]: Fixed.

**Pull request merged**